File: 0017-letter-combinations-of-a-phone-number/solution.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if digits == "":
            return []
        n_l = {"2":"abc", "3":"def", "4":"ghi", "5":"jkl", "6":"mno", "7":"pqrs","8":"tuv","9":"wxyz"}
        numbers = [2, 3, 4, 5, 6, 7, 8, 9]
        for number in numbers:
            if digits == str(number):
                p = n_l[digits]
                result = []
                for i in p:
                    result.append(i)
                return result
        else:
            r1 = self.letterCombinations(digits[:len(digits)//2])
            r2 = self.letterCombinations(digits[len(digits)//2:])
            result = []
            for i in range(len(r2)):
                for j in range(len(r1)):
                    result.append(r1[j] + r2[i])
            return result             
        
                
        

File: 0017-letter-combinations-of-a-phone-number/solution.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        
        vector<string> numberLetter = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        
        vector<string> result;
        
        for(int i = 0; i < digits.size(); i++)
        {
            int digitIndex = digits[i] - '0';
            vector<string> newWords;
            if(i == 0)
            {
                for(int k = 0; k < numberLetter[digitIndex].size(); k++)
                {
                    string newWord(1, numberLetter[digitIndex][k]);
                    
                    newWords.push_back(newWord);
                }
            }
            else
            {
                for(int k = 0; k < numberLetter[digitIndex].size(); k++)
                {
                    for(int j = 0; j < result.size(); j++)
                    {
                        string newWord = result[j] + numberLetter[digitIndex][k];
                        newWords.push_back(newWord);
                    }
                }
            }
            result = newWords;
        }
        
        return result;
        
    }
};
